### ISSUE TITLE
Dense Passage Retrieval for Open-Domain Question Answering

### ISSUE THREAD

**Issue by pocca2048** [pocca2048/ML-paper-reading#8]
**Conference** : EMNLP 2020
**Link** : https://arxiv.org/pdf/2004.04906.pdf
**Authors' Affiliation** : Facebook & UW
**TL;DR** : 제대로 학습하기만 하면 (pretrain 없이) 단순히 passage question pair로 파인튜닝만 해도 bm25를 이길수 있음

**Summary** :

open domain QA에서는 2스테이지로 진행됨. 1) context retriever가 몇개의 passage를 가져오고 2) machine reader가 거기서 정답을 고름.

여기서 context retriever 로 쓰이는게 보통 tf-idf나 bm25. 이것들은 일종의 sparse vector라고 볼수 있음. 이것과 반대되는게 dense vector.

완전히 다른 키워드로 쓰여있어도 의미를 인지하고 가깝게 매핑할 수 있다는 장점.

> A term-based system would have difficulty retrieving such a context, while a dense retrieval system would be able to better match “bad guy” with “villain” and fetch the correct context. Dense encodings are also learnable by adjusting the embedding functions, which provides additional flexibility to have a task-specific representation. With special in-memory data structures and indexing schemes, retrieval can be done efficiently using **maximum inner product search (MIPS) algorithms**

그러나 이런 dense embedding을 학습하는건 일반적으로 많은 labeled 데이터들이 필요하다고 믿어지고있음. 그래서 ORQA에서 나온 `Inverse cloze task` 이전에는 bm25를 이긴적이 없음. 여기서는 bm25를 이겼지만 단점 2가지가 있음: 

1. computationally intensive & 일반적인 문장이 목적함수에서 좋은 surrogate 이라는 보장이 없음.
2. context encoder는 finetune되지 않기 때문에 suboptimal함.

추가적인 pretraining없이 question / passage(or answers) pair만 가지고 dense embedding 모델을 잘 학습할 수 있을까? 이 논문에선 비교적 적은 q/p pair만 가지고 올바르게 학습하는 scheme에 집중한다.

Solution = the embedding is optimized for maximizing inner products of the question and relevant passage vectors, with an objective comparing all pairs of questions and passages in a batch.

컨트리뷰션: 1) 제대로 학습하기만 하면 단순히 passage question pair로 파인튜닝만 해도 bm25를 이길수 있음을 보임. 2) open domain qa에서 높은 retrieval precision은 높은 qa accuracy로 직결됨을 보임.

# 2 Background

# 3 Dense Passage Retriever (DPR)

이 논문의 초점은 open domain QA에서 retrieval을 개선하는 것임.

DPR = 그냥 SimCLR

BM25+DPR, using a linear combination of their scores as the new ranking function. Specifically, we obtain two initial sets of top-2000 passages based on BM25 and DPR, respectively, and rerank the union of them using BM25(q,p) + λ · sim(q, p) as the ranking function. We used λ = 1.1 based on the retrieval accuracy in the development set.

there is a high lexical overlap between passages and questions, which gives BM25 a clear advantage.

# 5 Experiments

## 5.2 Ablation Study on Model Training

Inbatch negative training (=simclr)가 1-of-N training 보다 성능이 조금이지만 좋음. 왜지?? 전혀 차이가 있을 이유가 없어보이는데. 정확히 1 of n이란게 뭘까?

batch size가 커질수록 성능이 좋아지는건 당연한 것으로 보이고.

마지막으로 제안되는게 흥미로운데, 일반적인 SimCLR에서 batch의 나머지 요소들을 negative로 간주하는것에 추가해서 hard negative를 추가하는 방식이다. 이 hard negative는 bm25로 뽑은 건데 이걸 넣어주고 안 넣어주고가 성능에 매우 큰 영향을 끼친다. 여기서도 배치사이즈가 커지면 성능은 좋아지고.

이 bm25로 추가된 negative들은 batch에 있는 다른 모든 문제에 대해서도 negative가 되는 것이다. bm25 negative를 하나만 추가한 것과 2개를 추가한 것에는 큰 차이를 보이지 않는다.

이걸 우리 데이터셋에 적용하려면 약간의 전처리가 필요할듯 하다 (ocr search request 별로 묶어줘야 함.)

dot product >= L2 >> cosine. triplet loss쓰나 NLL loss 쓰나 별 차이 없음.

### cross-dataset generalization

다른 데이터셋에서도 generalize를 잘 하는가? 잘한다. bm25 보다 훨씬 높다.

## 5.3 Qualitative Analysis

bm25는 lexical overlap이 있냐 없냐에 따라 크게 갈림. 또한 부분적으로 lexical overlap이 있으면서 다른 의미인 문장들을 걸러내지 못함. 반면 DPR은 lexical overlap이 없어도 잘하는 반면, lexical overlap이 중요한 경우 capture하지 못하는 경우가 있다. 즉, 각각이 장단점을 가진다. 둘다 적당히 결합해서 쓰면 좋을 것 같다.

## 5.4 Run-time Efficiency

> With the help of FAISS in-memory index for real-valued vectors , DPR can be made incredibly efficient, processing 995.0 questions per second, returning top 100 passages per question.

> In contrast, BM25/Lucene (implemented in Java, using file index) processes 23.7 questions per second per CPU thread.

반면 dense vector의 인덱스를 생성하는게 훨씬 오래걸린다. 21M 벡터에 대해서 FAISS에서 인덱스를 생성하는건 8.5시간이 걸리는 반면, 루씬에선 30분이면 된다.

# A Replication Study of Dense Passage Retriever

후속 논문 

https://arxiv.org/pdf/2104.05740.pdf 

원래 논문에서 bm25 가 잘못 평가되어있었음. bm25의 성능은 그것보다 훨씬 높음. 그래도 DPR이 bm25보다 성능이 높은건 사실.

bm25 + DPR hybrid 방식이 저평가되어있었음. bm25에 weight를 주는 방식으로 바꾸고 normalization까지 하면 성능이 확 오름. hybrid는 항상 DPR 단독이나 bm25 단독 보다 좋음.

retriever에서 나온 evidence를 scoring phase에 사용하면 end to end에서 성능이 더 오름.

by incorporating evidence from the retriever and an improved answer span scoring technique, we are able to improve end-to-end question answering effectiveness using exactly the same models as in the original work.
